src/ConnectionPool.jl

```diff
@@ -358,115 +383,45 @@ function purge(c::Connection)
     @ensure bytesavailable(c) == 0
 end
 
-"""
-The `pool` is a collection of open `Connection`s.  The `request`
-function calls `getconnection` to retrieve a connection from the
-`pool`.  When the `request` function has written a Request Message
-it calls `closewrite` to signal that the `Connection` can be reused
-for writing (to send the next Request). When the `request` function
-has read the Response Message it calls `closeread` to signal that
-the `Connection` can be reused for reading.
-"""
-const pool = [Vector{Connection}()]
-const poollock = [ReentrantLock()]
-const poolcondition = [Condition()]
-
 """
     closeall()
 
 Close all connections in `pool`.
 """
 function closeall()
-    lock(poollock[Threads.threadid()]) do
-        for c in pool[Threads.threadid()]
-            close(c)
+    for pod in values(POOL.conns)
+        while isready(pod.conns)
+            close(take!(pod.conns))
         end
-        empty!(pool[Threads.threadid()])
+        Threads.atomic_xchg!(pod.numactive, 0)
     end
-    notify(poolcondition[Threads.threadid()])
     return
 end
 
-"""
-    findwritable(type, host, port) -> Vector{Connection}
-
-Find `Connections` in the `pool` that are ready for writing.
-"""
-function findwritable(T::Type,
-                      host::AbstractString,
-                      port::AbstractString,
-                      pipeline_limit::Int,
-                      require_ssl_verification::Bool,
-                      reuse_limit::Int)
-
-    filter(c->(!c.writebusy &&
-               typeof(c.io) == T &&
-               c.host == host &&
-               c.port == port &&
-               c.pipeline_limit == pipeline_limit &&
-               c.require_ssl_verification == require_ssl_verification &&
-               c.writecount < reuse_limit &&
-               c.writecount - c.readcount < pipeline_limit + 1 &&
-               isopen(c.io)), pool[Threads.threadid()])
+struct Pod
+    conns::Channel{Connection}
+    numactive::Threads.Atomic{Int}
 end
 
-"""
-    findoverused(type, host, port, reuse_limit) -> Vector{Connection}
+Pod() = Pod(Channel{Connection}(Inf), Threads.Atomic{Int}())
 
-Find `Connections` in the `pool` that are over the reuse limit
-and have no more active readers.
-"""
-function findoverused(T::Type,
-                      host::AbstractString,
-                      port::AbstractString,
-                      reuse_limit::Int)
+decr!(pod::Pod) = Threads.atomic_sub!(pod.numactive, 1)
+incr!(pod::Pod) = Threads.atomic_add!(pod.numactive, 1)
 
-    filter(c->(typeof(c.io) == T &&
-               c.host == host &&
-               c.port == port &&
-               c.readcount >= reuse_limit &&
-               !c.readbusy &&
-               isopen(c.io)), pool[Threads.threadid()])
+function release(c::Connection)
+    pod = getpod(POOL, hashconn(c))
```

Comment:
You should never use `hash` as a dict key

Reply:
why? my thought here was avoiding the Tuple allocation of doing `(T, host, port, pipeline_limit, require_ssl_verification)` and having that be the pool Dict key.

Is there some danger w/ the core hash value changing during the lifetime of a session/application or something?

Reply:
Because of hash collisions?

Reply:
I mean, does it matter that you may have `hashconn(c1) == hashconn(c2)` for different `c1` and `c2`?

Reply:
I've added some docs to `hashconn`; yeah, I guess it's a bit of a misnomer using `hash` in the name, but what we're really checking here is whether Connections could be re-useable by multiple requests, so we're only interested in hashing a few key properties.